File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int len = temperatures.size();
        stack<int> st;
        vector<int> ngeIndices(len, -1);
        for (int i = 0; i < len; i++) {
            while (st.size() > 0 && temperatures[st.top()] < temperatures[i]) {
                int idx = st.top();
                st.pop();
                ngeIndices[idx] = i;
            }

            st.push(i);
        }

        vector<int> res(len, 0);

        for (int i = 0; i < len; i++) {
            if (ngeIndices[i] != -1) {
                res[i] = (ngeIndices[i] - i);
            }
        }

        return res;
    }
};

File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;

        int[] nge = new int[n];
        for (int i = 0; i < n; i++) {
            nge[i] = n;
        }

        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < n; i++) {
            int currTemp = temperatures[i];
            while (st.size() > 0 && currTemp > temperatures[st.peek()]) {
                int top = st.pop();
                nge[top] = i;
            }

            st.push(i);
        }

        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            if (nge[i] != n) {
                ans[i] = nge[i] - i;
            } else {
                nge[i] = 0;
            }
        }

        return ans;
    }
}